File: backend/app/services/accounting.py
"""
Accounting Service Module

This module provides business logic for accounting/session statistics
and reporting functionality, implementing comprehensive analytics and monitoring.
"""

from datetime import datetime, date, timedelta
from typing import Optional, List, Dict, Any
from decimal import Decimal

from app.repositories.accounting import (
    AccountingRepository,
    UserTrafficSummaryRepository,
    NasTrafficSummaryRepository
)
from app.schemas.accounting import (
    RadAcctResponse, RadAcctCreate, RadAcctUpdate,
    AccountingQuery, AccountingQueryFilters,
    SessionStatistics, TrafficStatistics, AccountingOverview,
    TopUsersReport, HourlyTrafficReport, DailyTrafficReport,
    NasUsageReport, CustomQueryResult, MaintenanceResult,
    PaginatedAccountingResponse, PaginatedTopUsersResponse,
    UserTrafficSummaryResponse, NasTrafficSummaryResponse,
    AccountingTimeRangeEnum
)
from app.core.exceptions import NotFoundError, ValidationError, BusinessLogicError
from app.core.logging import logger


class AccountingService:
    """Service for accounting operations and analytics"""

    def __init__(self, repository: AccountingRepository):
        self.repository = repository

    async def get_accounting_sessions(
        self,
        query: AccountingQuery
    ) -> PaginatedAccountingResponse:
        """Get paginated accounting sessions with filtering"""
        try:
            # Validate pagination parameters
            if query.page < 1:
                raise ValidationError("Page must be greater than 0")
            if query.page_size < 1 or query.page_size > 100:
                raise ValidationError("Page size must be between 1 and 100")

            # Process time range filters
            filters = self._process_filters(query.filters)

            # Get sessions from repository
            sessions, total = await self.repository.get_all_sessions(
                page=query.page,
                page_size=query.page_size,
                filters=filters,
                sort_field=query.sort_field,
                sort_order=query.sort_order
            )

            # Convert to response models
            session_responses = [self._to_response_model(
                session) for session in sessions]

            # Calculate pagination info
            total_pages = (total + query.page_size - 1) // query.page_size

            return PaginatedAccountingResponse(
                data=session_responses,
                total=total,
                page=query.page,
                page_size=query.page_size,
                total_pages=total_pages,
                has_next=query.page < total_pages,
                has_prev=query.page > 1
            )

        except Exception as e:
            logger.error(f"Error in get_accounting_sessions: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get accounting sessions: {str(e)}")

    async def get_session_by_id(self, session_id: int) -> RadAcctResponse:
        """Get accounting session by ID"""
        try:
            session = await self.repository.get_session_by_id(session_id)
            if not session:
                raise NotFoundError(f"Session with ID {session_id} not found")

            return self._to_response_model(session)

        except NotFoundError:
            raise
        except Exception as e:
            logger.error(f"Error in get_session_by_id: {str(e)}")
            raise BusinessLogicError(f"Failed to get session: {str(e)}")

    async def get_active_sessions(
        self,
        page: int = 1,
        page_size: int = 20,
        nas_ip: Optional[str] = None,
        username: Optional[str] = None
    ) -> PaginatedAccountingResponse:
        """Get active sessions with filtering"""
        try:
            # Validate pagination
            if page < 1:
                raise ValidationError("Page must be greater than 0")
            if page_size < 1 or page_size > 100:
                raise ValidationError("Page size must be between 1 and 100")

            # Get active sessions
            sessions, total = await self.repository.get_active_sessions(
                page=page,
                page_size=page_size,
                nas_ip=nas_ip,
                username=username
            )

            # Convert to response models
            session_responses = [self._to_response_model(
                session) for session in sessions]

            # Calculate pagination info
            total_pages = (total + page_size - 1) // page_size

            return PaginatedAccountingResponse(
                data=session_responses,
                total=total,
                page=page,
                page_size=page_size,
                total_pages=total_pages,
                has_next=page < total_pages,
                has_prev=page > 1
            )

        except Exception as e:
            logger.error(f"Error in get_active_sessions: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get active sessions: {str(e)}")

    async def get_user_sessions(
        self,
        username: str,
        page: int = 1,
        page_size: int = 20,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None
    ) -> PaginatedAccountingResponse:
        """Get sessions for a specific user"""
        try:
            # Validate input
            if not username:
                raise ValidationError("Username is required")

            # Get user sessions
            sessions, total = await self.repository.get_user_sessions(
                username=username,
                page=page,
                page_size=page_size,
                date_from=date_from,
                date_to=date_to
            )

            # Convert to response models
            session_responses = [self._to_response_model(
                session) for session in sessions]

            # Calculate pagination info
            total_pages = (total + page_size - 1) // page_size

            return PaginatedAccountingResponse(
                data=session_responses,
                total=total,
                page=page,
                page_size=page_size,
                total_pages=total_pages,
                has_next=page < total_pages,
                has_prev=page > 1
            )

        except Exception as e:
            logger.error(f"Error in get_user_sessions: {str(e)}")
            raise BusinessLogicError(f"Failed to get user sessions: {str(e)}")

    async def get_accounting_overview(
        self,
        time_range: Optional[AccountingTimeRangeEnum] = None,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None,
        filters: Optional[AccountingQueryFilters] = None
    ) -> AccountingOverview:
        """Get comprehensive accounting overview"""
        try:
            # Calculate date range
            if time_range and not date_from and not date_to:
                date_from, date_to = self._calculate_time_range(time_range)

            # Process filters
            filter_dict = self._process_filters(filters)
            filter_dict.update({
                'start_date': date_from,
                'end_date': date_to
            })

            # Get statistics
            stats = await self.repository.get_session_statistics(
                date_from=date_from,
                date_to=date_to,
                filters=filter_dict
            )

            # Create response models
            session_stats = SessionStatistics(
                total_sessions=stats['total_sessions'],
                active_sessions=stats['active_sessions'],
                completed_sessions=stats['completed_sessions'],
                average_session_duration=stats['average_session_duration'],
                total_session_time=stats['total_session_time'],
                unique_users=stats['unique_users']
            )

            traffic_stats = TrafficStatistics(
                total_input_octets=stats['total_input_octets'],
                total_output_octets=stats['total_output_octets'],
                total_bytes=stats['total_bytes'],
                total_input_packets=0,  # Not available in basic stats
                total_output_packets=0,  # Not available in basic stats
                total_packets=0,  # Not available in basic stats
                average_throughput=self._calculate_average_throughput(
                    stats['total_bytes'], stats['total_session_time']
                )
            )

            # Generate time period description
            time_period = self._format_time_period(
                time_range, date_from, date_to)

            return AccountingOverview(
                session_stats=session_stats,
                traffic_stats=traffic_stats,
                time_period=time_period,
                last_updated=datetime.utcnow()
            )

        except Exception as e:
            logger.error(f"Error in get_accounting_overview: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get accounting overview: {str(e)}")

    async def get_top_users_report(
        self,
        limit: int = 10,
        page: int = 1,
        page_size: int = 20,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None
    ) -> PaginatedTopUsersResponse:
        """Get top users by traffic consumption"""
        try:
            # Get top users data
            top_users_data = await self.repository.get_top_users_by_traffic(
                limit=limit * page,  # Get more data for pagination
                date_from=date_from,
                date_to=date_to
            )

            # Convert to response models
            top_users = []
            for user_data in top_users_data:
                top_users.append(TopUsersReport(
                    username=user_data['username'],
                    total_sessions=user_data['total_sessions'],
                    total_bytes=user_data['total_bytes'],
                    total_session_time=user_data['total_session_time'],
                    last_session=user_data['last_session'],
                    rank=user_data['rank']
                ))

            # Apply pagination
            start_idx = (page - 1) * page_size
            end_idx = start_idx + page_size
            paginated_users = top_users[start_idx:end_idx]

            total_pages = (len(top_users) + page_size - 1) // page_size

            return PaginatedTopUsersResponse(
                data=paginated_users,
                total=len(top_users),
                page=page,
                page_size=page_size,
                total_pages=total_pages
            )

        except Exception as e:
            logger.error(f"Error in get_top_users_report: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get top users report: {str(e)}")

    async def get_hourly_traffic_report(
        self,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None
    ) -> List[HourlyTrafficReport]:
        """Get hourly traffic distribution report"""
        try:
            hourly_data = await self.repository.get_hourly_traffic_distribution(
                date_from=date_from,
                date_to=date_to
            )

            # Convert to response models
            hourly_reports = []
            for data in hourly_data:
                hourly_reports.append(HourlyTrafficReport(
                    hour=data['hour'],
                    session_count=data['session_count'],
                    total_bytes=data['total_bytes'],
                    unique_users=data['unique_users']
                ))

            return hourly_reports

        except Exception as e:
            logger.error(f"Error in get_hourly_traffic_report: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get hourly traffic report: {str(e)}")

    async def get_nas_usage_report(
        self,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None
    ) -> List[NasUsageReport]:
        """Get NAS usage statistics report"""
        try:
            nas_data = await self.repository.get_nas_usage_statistics(
                date_from=date_from,
                date_to=date_to
            )

            # Convert to response models
            nas_reports = []
            for data in nas_data:
                nas_reports.append(NasUsageReport(
                    nasipaddress=data['nasipaddress'],
                    nas_name=None,  # Would need NAS repository to get name
                    total_sessions=data['total_sessions'],
                    active_sessions=data['active_sessions'],
                    total_bytes=data['total_bytes'],
                    utilization_percentage=data['utilization_percentage']
                ))

            return nas_reports

        except Exception as e:
            logger.error(f"Error in get_nas_usage_report: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get NAS usage report: {str(e)}")

    async def execute_custom_query(
        self,
        query_sql: str,
        parameters: Optional[Dict[str, Any]] = None,
        limit: int = 1000
    ) -> CustomQueryResult:
        """Execute custom accounting query"""
        try:
            # Validate query
            if not query_sql or not query_sql.strip():
                raise ValidationError("Query SQL is required")

            if limit < 1 or limit > 10000:
                raise ValidationError("Limit must be between 1 and 10000")

            # Execute query
            result = await self.repository.execute_custom_query(
                query_sql=query_sql,
                parameters=parameters,
                limit=limit
            )

            return CustomQueryResult(
                columns=result['columns'],
                rows=result['rows'],
                total_rows=result['total_rows'],
                execution_time=result['execution_time']
            )

        except Exception as e:
            logger.error(f"Error in execute_custom_query: {str(e)}")
            raise BusinessLogicError(
                f"Failed to execute custom query: {str(e)}")

    async def cleanup_old_sessions(
        self,
        days_old: int = 365,
        dry_run: bool = True
    ) -> MaintenanceResult:
        """Clean up old accounting sessions"""
        try:
            # Validate parameters
            if days_old < 30:
                raise ValidationError(
                    "Cannot delete sessions less than 30 days old")

            # Execute cleanup
            result = await self.repository.cleanup_old_sessions(
                days_old=days_old,
                dry_run=dry_run
            )

            return MaintenanceResult(
                operation_type=result['operation_type'],
                affected_rows=result['affected_rows'],
                execution_time=0.0,  # Repository doesn't track this
                success=result['success'],
                message=result['message']
            )

        except Exception as e:
            logger.error(f"Error in cleanup_old_sessions: {str(e)}")
            raise BusinessLogicError(
                f"Failed to cleanup old sessions: {str(e)}")

    # =====================================================================
    # Helper Methods
    # =====================================================================

    def _to_response_model(self, session) -> RadAcctResponse:
        """Convert database model to response model"""
        return RadAcctResponse(
            radacctid=session.radacctid,
            username=session.username,
            realm=session.realm,
            acctsessionid=session.acctsessionid,
            acctuniqueid=session.acctuniqueid,
            groupname=session.groupname,
            nasipaddress=str(
                session.nasipaddress) if session.nasipaddress else None,
            nasportid=session.nasportid,
            nasporttype=session.nasporttype,
            calledstationid=session.calledstationid,
            callingstationid=session.callingstationid,
            framedipaddress=str(
                session.framedipaddress) if session.framedipaddress else None,
            servicetype=session.servicetype,
            acctstarttime=session.acctstarttime,
            acctstoptime=session.acctstoptime,
            acctsessiontime=session.acctsessiontime,
            acctinputoctets=session.acctinputoctets or 0,
            acctoutputoctets=session.acctoutputoctets or 0,
            acctinputpackets=session.acctinputpackets or 0,
            acctoutputpackets=session.acctoutputpackets or 0,
            acctterminatecause=session.acctterminatecause,
            total_bytes=session.total_bytes,
            total_packets=session.total_packets,
            is_active=session.is_active,
            formatted_duration=self._format_duration(session.acctsessiontime)
        )

    def _process_filters(self, filters: Optional[AccountingQueryFilters]) -> Dict[str, Any]:
        """Process filters and convert to repository format"""
        if not filters:
            return {}

        filter_dict = {}

        # Direct field mappings
        for field in ['username', 'groupname', 'nasipaddress', 'framedipaddress',
                      'callingstationid', 'servicetype']:
            value = getattr(filters, field, None)
            if value:
                filter_dict[field] = value

        # Date range handling
        if filters.time_range and not filters.start_date and not filters.end_date:
            start_date, end_date = self._calculate_time_range(
                filters.time_range)
            filter_dict['start_date'] = start_date
            filter_dict['end_date'] = end_date
        else:
            if filters.start_date:
                filter_dict['start_date'] = filters.start_date
            if filters.end_date:
                filter_dict['end_date'] = filters.end_date

        # Status filters
        if filters.status:
            filter_dict['status'] = filters.status
        if filters.active_only:
            filter_dict['active_only'] = filters.active_only

        # Traffic filters
        for field in ['min_input_octets', 'max_input_octets', 'min_output_octets',
                      'max_output_octets', 'min_session_time', 'max_session_time']:
            value = getattr(filters, field, None)
            if value is not None:
                filter_dict[field] = value

        return filter_dict

    def _calculate_time_range(self, time_range: AccountingTimeRangeEnum) -> tuple[datetime, datetime]:
        """Calculate start and end dates for predefined time ranges"""
        now = datetime.now()
        today = now.replace(hour=0, minute=0, second=0, microsecond=0)

        if time_range == AccountingTimeRangeEnum.TODAY:
            return today, now
        elif time_range == AccountingTimeRangeEnum.YESTERDAY:
            yesterday = today - timedelta(days=1)
            return yesterday, today
        elif time_range == AccountingTimeRangeEnum.THIS_WEEK:
            week_start = today - timedelta(days=today.weekday())
            return week_start, now
        elif time_range == AccountingTimeRangeEnum.LAST_WEEK:
            week_start = today - timedelta(days=today.weekday() + 7)
            week_end = week_start + timedelta(days=7)
            return week_start, week_end
        elif time_range == AccountingTimeRangeEnum.THIS_MONTH:
            month_start = today.replace(day=1)
            return month_start, now
        elif time_range == AccountingTimeRangeEnum.LAST_MONTH:
            if today.month == 1:
                last_month_start = today.replace(
                    year=today.year-1, month=12, day=1)
            else:
                last_month_start = today.replace(month=today.month-1, day=1)

            current_month_start = today.replace(day=1)
            return last_month_start, current_month_start
        elif time_range == AccountingTimeRangeEnum.THIS_YEAR:
            year_start = today.replace(month=1, day=1)
            return year_start, now
        else:
            # Default to last 30 days
            return today - timedelta(days=30), now

    def _calculate_average_throughput(self, total_bytes: int, total_session_time: int) -> Optional[float]:
        """Calculate average throughput in bytes per second"""
        if total_session_time and total_session_time > 0:
            return float(total_bytes) / float(total_session_time)
        return None

    def _format_time_period(
        self,
        time_range: Optional[AccountingTimeRangeEnum],
        date_from: Optional[datetime],
        date_to: Optional[datetime]
    ) -> str:
        """Format time period description"""
        if time_range:
            return time_range.value.replace('_', ' ').title()
        elif date_from and date_to:
            return f"{date_from.strftime('%Y-%m-%d')} to {date_to.strftime('%Y-%m-%d')}"
        elif date_from:
            return f"From {date_from.strftime('%Y-%m-%d')}"
        elif date_to:
            return f"Until {date_to.strftime('%Y-%m-%d')}"
        else:
            return "All time"

    def _format_duration(self, seconds: Optional[int]) -> Optional[str]:
        """Format duration in human readable format"""
        if not seconds:
            return None

        hours = seconds // 3600
        minutes = (seconds % 3600) // 60
        secs = seconds % 60

        if hours > 0:
            return f"{hours}h {minutes}m {secs}s"
        elif minutes > 0:
            return f"{minutes}m {secs}s"
        else:
            return f"{secs}s"


# =====================================================================
# User Traffic Summary Service
# =====================================================================

class UserTrafficSummaryService:
    """Service for user traffic summary operations"""

    def __init__(self, repository: UserTrafficSummaryRepository):
        self.repository = repository

    async def get_user_traffic_summary(
        self,
        username: str,
        date_from: Optional[date] = None,
        date_to: Optional[date] = None
    ) -> List[UserTrafficSummaryResponse]:
        """Get traffic summary for a user"""
        try:
            summaries = await self.repository.get_user_summary(
                username=username,
                date_from=date_from,
                date_to=date_to
            )

            return [
                UserTrafficSummaryResponse.from_orm(summary)
                for summary in summaries
            ]

        except Exception as e:
            logger.error(f"Error getting user traffic summary: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get user traffic summary: {str(e)}")


# =====================================================================
# NAS Traffic Summary Service
# =====================================================================

class NasTrafficSummaryService:
    """Service for NAS traffic summary operations"""

    def __init__(self, repository: NasTrafficSummaryRepository):
        self.repository = repository

    async def get_nas_traffic_summary(
        self,
        nasipaddress: str,
        date_from: Optional[date] = None,
        date_to: Optional[date] = None
    ) -> List[NasTrafficSummaryResponse]:
        """Get traffic summary for a NAS"""
        try:
            summaries = await self.repository.get_nas_summary(
                nasipaddress=nasipaddress,
                date_from=date_from,
                date_to=date_to
            )

            return [
                NasTrafficSummaryResponse.from_orm(summary)
                for summary in summaries
            ]

        except Exception as e:
            logger.error(f"Error getting NAS traffic summary: {str(e)}")
            raise BusinessLogicError(
                f"Failed to get NAS traffic summary: {str(e)}")


# Export services
__all__ = [
    "AccountingService",
    "UserTrafficSummaryService",
    "NasTrafficSummaryService",
]

File: frontend/src/types/accounting.ts
/**
 * Accounting Types
 * 
 * TypeScript type definitions for accounting/session statistics module
 */

// =====================================================================
// Base Types
// =====================================================================

export interface AccountingSession {
  radacctid: number
  username: string
  realm?: string
  acctsessionid: string
  acctuniqueid?: string
  groupname?: string
  nasipaddress: string
  nasportid?: string
  nasporttype?: string
  calledstationid?: string
  callingstationid?: string
  framedipaddress?: string
  servicetype?: string
  acctstarttime?: string
  acctstoptime?: string
  acctsessiontime?: number
  acctinputoctets?: number
  acctoutputoctets?: number
  acctinputpackets?: number
  acctoutputpackets?: number
  acctterminatecause?: string
  total_bytes: number
  total_packets: number
  is_active: boolean
  formatted_duration?: string
}

export interface UserTrafficSummary {
  id: number
  username: string
  summary_date: string
  total_sessions: number
  total_session_time: number
  total_input_octets: number
  total_output_octets: number
  total_bytes: number
  avg_session_duration: number
  created_at: string
  updated_at: string
}

export interface NasTrafficSummary {
  id: number
  nasipaddress: string
  summary_date: string
  total_sessions: number
  total_session_time: number
  total_input_octets: number
  total_output_octets: number
  total_bytes: number
  unique_users: number
  avg_session_duration: number
  created_at: string
  updated_at: string
}

// =====================================================================
// Filter Types
// =====================================================================

export type AccountingTimeRangeEnum = 
  | 'TODAY'
  | 'YESTERDAY' 
  | 'THIS_WEEK'
  | 'LAST_WEEK'
  | 'THIS_MONTH'
  | 'LAST_MONTH'
  | 'THIS_YEAR'

export type AccountingStatusEnum = 'active' | 'completed' | 'all'

export interface AccountingQueryFilters {
  username?: string | null
  groupname?: string | null
  nasipaddress?: string | null
  framedipaddress?: string | null
  callingstationid?: string | null
  servicetype?: string | null
  time_range?: AccountingTimeRangeEnum | null
  start_date?: string | null
  end_date?: string | null
  status?: AccountingStatusEnum | null
  active_only?: boolean
  min_input_octets?: number | null
  max_input_octets?: number | null
  min_output_octets?: number | null
  max_output_octets?: number | null
  min_session_time?: number | null
  max_session_time?: number | null
}

export interface SessionFilters extends AccountingQueryFilters {}

export interface ActiveSessionFilters {
  username?: string | null
  nas_ip?: string | null
  sort_by?: string
}

export interface ReportFilters {
  time_range?: AccountingTimeRangeEnum | null
  start_date?: string | null
  end_date?: string | null
  report_type?: 'summary' | 'detailed' | 'comparative'
}

// =====================================================================
// Query Types
// =====================================================================

export interface AccountingQuery {
  page: number
  page_size: number
  sort_field?: string
  sort_order?: 'asc' | 'desc'
  filters?: AccountingQueryFilters
}

export interface CustomQueryRequest {
  query_sql: string
  parameters?: Record<string, any>
  limit?: number
}

export interface MaintenanceRequest {
  days_old: number
  dry_run?: boolean
}

// =====================================================================
// Statistics Types
// =====================================================================

export interface SessionStatistics {
  total_sessions: number
  active_sessions: number
  completed_sessions: number
  average_session_duration: number
  total_session_time: number
  unique_users: number
}

export interface TrafficStatistics {
  total_input_octets: number
  total_output_octets: number
  total_bytes: number
  total_input_packets: number
  total_output_packets: number
  total_packets: number
  average_throughput?: number
}

export interface ActiveSessionStatistics {
  total_active: number
  total_current_traffic: number
  unique_users: number
  average_session_duration: number
}

// =====================================================================
// Report Types
// =====================================================================

export interface AccountingOverview {
  session_stats: SessionStatistics
  traffic_stats: TrafficStatistics
  time_period: string
  last_updated: string
}

export interface TopUsersReport {
  username: string
  total_sessions: number
  total_bytes: number
  total_session_time: number
  last_session?: string
  rank: number
}

export interface HourlyTrafficReport {
  hour: number
  session_count: number
  total_bytes: number
  unique_users: number
}

export interface DailyTrafficReport {
  date: string
  session_count: number
  total_bytes: number
  unique_users: number
  peak_hour: number
}

export interface NasUsageReport {
  nasipaddress: string
  nas_name?: string
  total_sessions: number
  active_sessions: number
  total_bytes: number
  utilization_percentage: number
}

export interface CustomQueryResult {
  columns: string[]
  rows: any[][]
  total_rows: number
  execution_time: number
}

export interface MaintenanceResult {
  operation_type: string
  affected_rows: number
  execution_time: number
  success: boolean
  message: string
}

// =====================================================================
// Response Types
// =====================================================================

export interface PaginatedAccountingResponse {
  data: AccountingSession[]
  total: number
  page: number
  page_size: number
  total_pages: number
  has_next: boolean
  has_prev: boolean
}

export interface PaginatedTopUsersResponse {
  data: TopUsersReport[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface UserTrafficSummaryResponse extends UserTrafficSummary {}
export interface NasTrafficSummaryResponse extends NasTrafficSummary {}

// =====================================================================
// Store State Types
// =====================================================================

export interface AccountingState {
  // Session data
  sessions: AccountingSession[]
  activeSessions: AccountingSession[]
  recentSessions: AccountingSession[]
  
  // Pagination
  sessionsPagination: {
    page: number
    page_size: number
    total: number
    total_pages: number
    has_next: boolean
    has_prev: boolean
  }
  
  activeSessionsPagination: {
    page: number
    page_size: number
    total: number
    total_pages: number
    has_next: boolean
    has_prev: boolean
  }
  
  // Statistics
  overview: AccountingOverview | null
  sessionStatistics: SessionStatistics
  activeSessionStatistics: ActiveSessionStatistics
  
  // Reports
  topUsers: TopUsersReport[]
  overviewReport: AccountingOverview | null
  topUsersReport: TopUsersReport[]
  trafficAnalysisReport: any | null
  nasUsageReport: NasUsageReport[]
  timeAnalysisReport: any | null
  
  // Traffic summaries
  userTrafficSummaries: UserTrafficSummary[]
  nasTrafficSummaries: NasTrafficSummary[]
  
  // Loading states
  loading: {
    sessions: boolean
    activeSessions: boolean
    overview: boolean
    reports: boolean
    export: boolean
  }
  
  // Error states
  error: string | null
}

// =====================================================================
// API Service Types
// =====================================================================

export interface AccountingApiParams {
  sessions?: AccountingQuery
  activeSessions?: {
    page?: number
    page_size?: number
    nas_ip?: string
    username?: string
  }
  overview?: {
    time_range?: AccountingTimeRangeEnum
    date_from?: string
    date_to?: string
    filters?: AccountingQueryFilters
  }
  topUsers?: {
    limit?: number
    page?: number
    page_size?: number
    date_from?: string
    date_to?: string
  }
  reports?: ReportFilters
  export?: {
    type: string
    format: string
    filters?: any
  }
}

// =====================================================================
// Component Props Types
// =====================================================================

export interface SessionsListProps {
  sessions: AccountingSession[]
  loading?: boolean
  compact?: boolean
  showActions?: boolean
}

export interface SessionDetailsDialogProps {
  modelValue: boolean
  session: AccountingSession | null
}

export interface TopUsersListProps {
  users: TopUsersReport[]
  loading?: boolean
  compact?: boolean
  showActions?: boolean
}

export interface OverviewReportProps {
  data: AccountingOverview | null
  loading?: boolean
  filters: ReportFilters
}

// =====================================================================
// Utility Types
// =====================================================================

export type SortOrder = 'asc' | 'desc'

export interface TableHeader {
  title: string
  key: string
  sortable?: boolean
  width?: number
  align?: 'start' | 'center' | 'end'
}

export interface ChartDataPoint {
  label: string
  value: number
  color?: string
}

export interface ExportOptions {
  format: 'csv' | 'xlsx' | 'pdf'
  filename?: string
  includeCharts?: boolean
}

// =====================================================================
// Event Types
// =====================================================================

export interface SessionEvent {
  type: 'view' | 'edit' | 'delete' | 'disconnect' | 'terminate'
  session: AccountingSession
}

export interface ReportEvent {
  type: 'export' | 'refresh' | 'filter'
  reportType: string
  data?: any
}

// =====================================================================
// Form Types
// =====================================================================

export interface SessionFilterForm {
  username: string
  nasipaddress: string
  dateRange: string[]
  status: AccountingStatusEnum
  trafficRange: [number, number]
}

export interface ReportConfigForm {
  timeRange: AccountingTimeRangeEnum
  reportTypes: string[]
  exportFormat: 'csv' | 'xlsx' | 'pdf'
  includeCharts: boolean
}

// =====================================================================
// Validation Types
// =====================================================================

export interface ValidationError {
  field: string
  message: string
}

export interface ValidationResult {
  valid: boolean
  errors: ValidationError[]
}

// =====================================================================
// Theme Types
// =====================================================================

export interface AccountingTheme {
  primary: string
  secondary: string
  success: string
  warning: string
  error: string
  info: string
}

// Note: All types are already exported as named exports above.
// No need for additional export declarations.